src/targets/maven.ts

```diff
@@ -258,12 +261,95 @@ export class MavenTarget extends BaseTarget {
    * @param distDir directory of the distribution.
    */
   private async uploadDistribution(distDir: string): Promise<void> {
+    const bomFile = await this.getBomFileInDist(distDir);
+    if (bomFile) {
+      this.logger.debug('Found BOM: ', bomFile);
+      await this.uploadBomDistribution(bomFile);
+    } else {
+      this.logger.debug('Did not find a BOM.');
+      await this.uploadPomDistribution(distDir);
+    }
+  }
+
+  /**
+   * Returns the path to the BOM file in the given distribution directory, and
+   * `undefined` if there isn't any.
+   */
+  private async getBomFileInDist(distDir: string): Promise<string | undefined> {
+    const pomFilepath = join(distDir, POM_DEFAULT_FILENAME);
+    if (await this.isBomFile(pomFilepath)) {
+      return pomFilepath;
+    }
+
+    // There may be several files in the ZIP-ed artifact with the same name,
+    // where the BOM may be one of them (there may not be a BOM). Files may be
+    // renamed when extracting the ZIP, so the default name (`pom-default.xml`)
+    // may not match. It's assumed that any renaming keeps the same extension,
+    // so all files with the same extension are checked to identify the BOM.
+    // TODO: make sure all scenarios are considered and tested.
+    // Each file system may handle this case differently, and attended vs
+    // unattended mode also have different behaviours. It's not desired to get
+    // the BOM renamed in such a way that isn't handled by the target.
+    const filesInDir = await fsPromises.readdir(distDir);
+    const potentialPoms = filesInDir
+      .filter(f => extname(f) === POM_FILE_EXT)
+      .filter(f => f !== POM_DEFAULT_FILENAME)
+      .map(f => join(distDir, f));
+
+    for (const f of potentialPoms) {
+      if (await this.isBomFile(f)) {
+        return f;
+      }
+    }
+
+    return undefined;
+  }
+
+  /**
+   * Returns whether the given POM is a BOM.
+   *
+   * A BOM file is a POM file with the following key:
+   * `<packaging>pom</packaging>`, usually named as `pom-default.xml`.
+   *
+   * @param pomFilepath path to the POM.
+   * @returns true if the POM is a BOM.
+   */
+  public async isBomFile(pomFilepath: string): Promise<boolean> {
+    try {
+      await fsPromises.access(pomFilepath, fsConstants.R_OK);
```

Comment:
Why do we check this explicitly? `fsPromises.readFile` would throw an exception if we cannot read the file anyway?